File: mycelium-cli/src/mycelium/board/upstream.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Mycelium Contributors

"""Put a status provider's answer on the row that mentioned it.

The hub resolves every external reference a room's text names and returns them
keyed by the board row ids that mention them, so this does no parsing: it looks
up a row's id and lands the answer under ``upstream``.

**The state is the field; the rest are companions.** ``upstream`` carries the
state alone — a closed vocabulary the board can group, filter and color by,
the same way it treats ``status`` and ``ci``. The provider's own wording, its
link and the answer's age ride alongside in their own fields, the same way a
row already carries ``pr``, ``ci`` and ``branch``.

**A row that names two things shows the worse one.** A task blocked on one
pull request and green on another reads as blocked; ``upstream_count`` says
there was more than one.

**Not knowing yet is ``upstream_pending``, not ``unknown``.** The hub answers a
read from cache and refreshes behind it, so the first read of a room has
references but no answers yet; ``unknown`` is reserved for a state a provider
reports when it meets something it cannot place. A pending row carries no
``upstream`` at all, so grouping by the field never collects a bucket of rows
that were merely early.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from mycelium.board.model import LiveItem

#: Worst-first ranking: a row with several references takes the first state in
#: this order that any of them is in.
UPSTREAM_STATES = ["failed", "blocked", "pending", "ok", "done", "unknown"]

FIELD = "upstream"


def _rank(state: str | None) -> int:
    try:
        return UPSTREAM_STATES.index(state or "unknown")
    except ValueError:
        return len(UPSTREAM_STATES)


def _age(seconds: float | None) -> str | None:
    """A human reading of how old an answer is, in the same shape a row's other
    ages take. An answer with no age was never fetched, and says nothing."""
    if seconds is None:
        return None
    minutes = int(seconds // 60)
    if minutes < 1:
        return "just now"
    if minutes < 60:
        return f"{minutes}m ago"
    hours = minutes // 60
    return f"{hours}h ago" if hours < 24 else f"{hours // 24}d ago"


def attach_upstream(items: list[LiveItem], status: dict[str, Any] | None) -> list[LiveItem]:
    """Land each row's upstream answer on it, in place, and return the rows.

    A row nothing was found for is left exactly as it was: no field, rather than
    an empty one, so a view can tell "nothing upstream" from "upstream unknown".
    """
    if not status:
        return items

    by_ref = {entry.get("ref"): entry for entry in status.get("refs", []) if entry.get("ref")}
    rows = status.get("rows") or {}

    refreshing = bool(status.get("refreshing"))

    for item in items:
        answers = [by_ref[ref] for ref in rows.get(item.id, []) if ref in by_ref]
        if not answers:
            continue
        # A resolved answer outranks one still coming, so a row with two
        # references shows what is known rather than waiting on the slowest.
        answers.sort(key=lambda a: (a.get("state") is None, _rank(a.get("state"))))
        worst = answers[0]

        if len(answers) > 1:
            item.fields["upstream_count"] = len(answers)
        item.fields["upstream_freshness"] = worst.get("freshness")

        # Nothing has come back for this row yet. Say so, rather than showing a
        # state nobody reported.
        if worst.get("state") is None:
            item.fields["upstream_pending"] = worst.get("freshness") == "missing" or refreshing
            item.fields["upstream_label"] = worst.get("error")
            continue

        item.fields[FIELD] = worst["state"]
        # An errored answer has no label of its own, so the reason stands in:
        # "not visible to this token" is the honest wording for that row.
        item.fields["upstream_label"] = worst.get("label") or worst.get("error")
        item.fields["upstream_url"] = worst.get("url")
        item.fields["upstream_age"] = _age(worst.get("age_seconds"))
    return items

File: mycelium-frontend/src/lib/board/upstream.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Mycelium Contributors

/**
 * Put a status provider's answer on the row that mentioned it.
 *
 * The hub resolves every external reference a room's text names and returns
 * them keyed by the board row ids that mention them, so this does no parsing:
 * it looks up a row's id and lands the answer under `upstream`.
 *
 * **The state is the field; the rest are companions.** `upstream` carries the
 * state alone (a closed vocabulary the board can group, filter and colour by,
 * the same way it treats `status` and `ci`) rather than an object, which
 * inference would type as free text and no view could group. The provider's own
 * wording, its link and the answer's age ride alongside in their own fields,
 * which is how a row already carries `pr`, `ci` and `branch`.
 *
 * **A row that names two things shows the worse one.** Rows are how a board says
 * what needs you, so a task blocked on one pull request and green on another
 * reads as blocked; `upstream_count` says there was more than one, so the
 * summary never quietly stands in for the whole.
 *
 * **Not knowing yet is not a state.** The hub answers a read from cache and
 * refreshes behind it, so the first read of a room has references but no answers
 * for them. That is `upstream_pending`, and deliberately not `unknown`, which is
 * a real thing a provider says when it meets a state it cannot place. A row that
 * is pending carries no `upstream` at all, so a board grouping by the field
 * never collects a bucket of rows that were merely early.
 */

import type { LiveItem } from "./item";

/**
 * Worst first. A row with several references takes the first state in this
 * order that any of them is in, because a board exists to surface the thing
 * that needs a person, not to average.
 */
export const UPSTREAM_STATES = ["failed", "blocked", "pending", "ok", "done", "unknown"] as const;

export type UpstreamState = (typeof UPSTREAM_STATES)[number];

export const UPSTREAM_FIELD = "upstream";

/** One resolved reference, as `GET /rooms/{room}/status` returns it. */
export interface UpstreamRef {
  ref: string;
  provider: string;
  kind: string;
  id: string;
  url: string | null;
  freshness: "fresh" | "stale" | "missing" | "error";
  state: UpstreamState | null;
  label: string | null;
  age_seconds: number | null;
  error: string | null;
  origins: string[];
}

export interface RoomStatus {
  room: string;
  field: string;
  providers: string[];
  refs: UpstreamRef[];
  rows: Record<string, string[]>;
  refreshing: boolean;
}

function rank(state: string | null | undefined): number {
  const i = (UPSTREAM_STATES as readonly string[]).indexOf(state ?? "unknown");
  return i === -1 ? UPSTREAM_STATES.length : i;
}

/**
 * A human reading of how old an answer is, in the same shape a row's other ages
 * take. An answer with no age was never fetched, and says nothing.
 */
export function upstreamAge(seconds: number | null | undefined): string | null {
  if (seconds === null || seconds === undefined) return null;
  const minutes = Math.floor(seconds / 60);
  if (minutes < 1) return "just now";
  if (minutes < 60) return `${minutes}m ago`;
  const hours = Math.floor(minutes / 60);
  return hours < 24 ? `${hours}h ago` : `${Math.floor(hours / 24)}d ago`;
}

/**
 * Land each row's upstream answer on it, returning new rows.
 *
 * A row nothing was found for is returned untouched: no field, rather than an
 * empty one, so a view can tell "nothing upstream" from "upstream unknown".
 */
export function attachUpstream(items: LiveItem[], status: RoomStatus | null | undefined): LiveItem[] {
  if (!status) return items;

  const byRef = new Map(status.refs.map(r => [r.ref, r]));
  const rows = status.rows ?? {};

  return items.map(item => {
    const answers = (rows[item.id] ?? [])
      .map(ref => byRef.get(ref))
      .filter((r): r is UpstreamRef => r !== undefined);
    if (answers.length === 0) return item;

    // A resolved answer outranks one still coming, so a row with two references
    // shows what is known rather than waiting on the slowest.
    answers.sort((a, b) => {
      const known = Number(b.state !== null) - Number(a.state !== null);
      return known !== 0 ? known : rank(a.state) - rank(b.state);
    });
    const worst = answers[0];

    // Nothing has come back for this row yet. Say so, rather than showing a
    // state nobody reported.
    if (worst.state === null) {
      return {
        ...item,
        fields: {
          ...item.fields,
          upstream_pending: worst.freshness === "missing" || status.refreshing,
          upstream_label: worst.error,
          upstream_freshness: worst.freshness,
          ...(answers.length > 1 ? { upstream_count: answers.length } : {}),
        },
      };
    }

    return {
      ...item,
      fields: {
        ...item.fields,
        [UPSTREAM_FIELD]: worst.state,
        // An errored answer has no label of its own, so the reason stands in:
        // "not visible to this token" is the honest wording for that row.
        upstream_label: worst.label ?? worst.error,
        upstream_url: worst.url,
        upstream_age: upstreamAge(worst.age_seconds),
        upstream_freshness: worst.freshness,
        ...(answers.length > 1 ? { upstream_count: answers.length } : {}),
      },
    };
  });
}
